File: tests-csharp/Pages/CheckoutOverviewPage.cs
using Microsoft.Playwright;
using Tests.CSharp.Pages;

namespace Tests.CSharp.Pages;

public class CheckoutOverviewPage : BasePage
{
    private readonly ILocator _finishButton;
    private readonly ILocator _cancelButton;
    private readonly ILocator _cartItems;
    private readonly ILocator _subtotalLabel;
    private readonly ILocator _taxLabel;
    private readonly ILocator _totalLabel;

    public CheckoutOverviewPage(IPage page) : base(page)
    {
        _finishButton = Page.Locator("[data-test='finish']");
        _cancelButton = Page.Locator("[data-test='cancel']");
        _cartItems = Page.Locator(".cart_item");
        _subtotalLabel = Page.Locator(".summary_subtotal_label");
        _taxLabel = Page.Locator(".summary_tax_label");
        _totalLabel = Page.Locator(".summary_total_label");
    }

    public async Task FinishOrderAsync()
    {
        try
        {
            await WaitForElementAsync(_finishButton, 15000);
            await ClickElementAsync(_finishButton);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Finish button not found or not clickable: {ex}");
            throw;
        }
    }

    public async Task CancelAsync()
    {
        await ClickElementAsync(_cancelButton);
    }

    public async Task<int> GetCartItemsCountAsync()
    {
        return await _cartItems.CountAsync();
    }

    public async Task<string> GetSubtotalAsync()
    {
        return await GetElementTextAsync(_subtotalLabel);
    }

    public async Task<string> GetTaxAsync()
    {
        return await GetElementTextAsync(_taxLabel);
    }

    public async Task<string> GetTotalAsync()
    {
        return await GetElementTextAsync(_totalLabel);
    }
}

File: tests-playwright/pages/checkout-overview-page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base-page';

export class CheckoutOverviewPage extends BasePage {
  private readonly overviewContainer: Locator;
  private readonly cartItems: Locator;
  private readonly subtotalLabel: Locator;
  private readonly taxLabel: Locator;
  private readonly totalLabel: Locator;
  private readonly finishButton: Locator;
  private readonly cancelButton: Locator;

  constructor(page: Page) {
    super(page);
    this.overviewContainer = page.locator('#checkout_summary_container');
    this.cartItems = page.locator('.cart_item');
    this.subtotalLabel = page.locator('.summary_subtotal_label');
    this.taxLabel = page.locator('.summary_tax_label');
    this.totalLabel = page.locator('.summary_total_label');
    this.finishButton = page.locator('[data-test="finish"]');
    this.cancelButton = page.locator('[data-test="cancel"]');
  }

  async getCartItemsCount(): Promise<number> {
    return await this.cartItems.count();
  }

  async getItemTitle(index: number): Promise<string> {
    const item = this.cartItems.nth(index);
    const titleElement = item.locator('.inventory_item_name');
    return await this.getElementText(titleElement);
  }

  async getItemPrice(index: number): Promise<string> {
    const item = this.cartItems.nth(index);
    const priceElement = item.locator('.inventory_item_price');
    return await this.getElementText(priceElement);
  }

  async getSubtotal(): Promise<string> {
    return await this.getElementText(this.subtotalLabel);
  }

  async getTax(): Promise<string> {
    return await this.getElementText(this.taxLabel);
  }

  async getTotal(): Promise<string> {
    return await this.getElementText(this.totalLabel);
  }

  async finishOrder(): Promise<void> {
    try {
      await this.waitForElement(this.finishButton, 15000);
      await this.clickElement(this.finishButton);
    } catch (error) {
      console.log('Finish button not found or not clickable:', error);
      throw error;
    }
  }

  async cancelOrder(): Promise<void> {
    await this.clickElement(this.cancelButton);
  }

  async isOverviewPageLoaded(): Promise<boolean> {
    return await this.isElementVisible(this.overviewContainer);
  }
}
